crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs

```diff
@@ -0,0 +1,237 @@
+
+/// ## What it does
+///
+/// Checks for use of `TypeParam`, `TypeParamTuple`, and `ParamSpec` annotations on generic
+/// functions and classes.
+///
+/// ## Why is this bad?
+///
+/// Special type parameter syntax was introduced in Python 3.12 by [PEP 695] for defining generic
+/// functions and classes. This syntax is easier to read and provides cleaner support for generics.
+///
+/// ## Known problems
+///
+/// [PEP 695] uses inferred variance for type parameters, instead of the `covariant` and
+/// `contravariant` keywords used by `TypeParam` variables. As such, rewriting a `TypeParam`
+/// variable to an in-line type parameter may change its variance.
+///
+/// Unlike `TypeParam` variables, [PEP 695]-style type parameters cannot be used at runtime. For
+/// example, calling `isinstance(x, T)` with type parameter `T` will raise a `TypeError`. As such,
+/// rewriting a `TypeParam` as a type parameter will cause issues for parameters that are used for
+/// such runtime checks.
+///
+/// ## Example
+/// ```python
+/// T = TypeVar("T")
+/// class GenericClass(Generic[T]):
+///     ...
+/// ```
+///
+/// Use instead:
+/// ```python
+/// class GenericClass[T]:
+///     ...
+/// ```
+///
+/// [PEP 695]: https://peps.python.org/pep-0695/
+#[derive(ViolationMetadata)]
+pub(crate) struct NonPEP695TypeParameter {
+    name: String,
+    generic_kind: GenericKind,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum GenericKind {
+    GenericClass,
+    GenericFunction,
+}
+
+impl Violation for NonPEP695TypeParameter {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Always;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let NonPEP695TypeParameter { name, generic_kind } = self;
+        match generic_kind {
+            GenericKind::GenericClass => {
+                format!("Generic class `{name}` uses `Generic` subclass instead of type parameters")
+            }
+            GenericKind::GenericFunction => {
+                format!("Generic function `{name}` should use type parameters")
+            }
+        }
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Use type parameters".to_string())
+    }
+}
+
+/// UP046
+pub(crate) fn non_pep695_generic_class(checker: &mut Checker, class_def: &StmtClassDef) {
+    if checker.settings.target_version < PythonVersion::Py312 {
+        return;
+    }
+
+    let StmtClassDef {
+        name,
+        type_params,
+        arguments,
+        ..
+    } = class_def;
+
+    // it's a runtime error to mix type_params and Generic, so bail out early if we see existing
+    // type_params
+    if type_params.is_some() {
+        return;
+    }
+
+    let Some(arguments) = arguments.as_ref() else {
+        return;
+    };
+
+    // TODO(brent) only accept a single, Generic argument for now. I think it should be fine to have
+    // other arguments, but this simplifies the fix just to delete the argument list for now
```

Comment:
I support this -- it's good to cut scope for the first version of the rule. Could you possibly add a note to the rule's documentation explaining this limitation for now, though?